### ISSUE TITLE
Android 0.4.17.x: When returning from lock screen game closes to phone interface

### ISSUE THREAD

**Issue by juozaspo** [minetest/minetest#7501]
##### Issue type

 - Bug report
 
##### Minetest version

Google Play Store build of minetest `0.4.17.x`, version `0.4.17-38b7f3d8-Android`, apk version `0.4.17.20`, based on detached commit 38b7f3d8 not in source tree of branch `stable-0.4`, hashes diverge at ea86ecf015cd82d2daa4c0f4a82096402290a5a6
##### OS / Hardware

Operating system: Android 7.0
CPU: HiSilicon Kirin 650 2.11 GHz (arm 8 cores)

GPU model: ARM Mali-T830
OpenGL version: 3.2

##### Summary

While running minetest, both ingame and main menu, entering lock screen and returning from it causes the game to close to phone interface. 
Adding `debug_log_level = verbose` shows interesting results in log file.
Verbose `debug.txt` (while in main menu): 
```
-------------
  Separator
-------------

2018-06-29 10:00:01: VERBOSE[Main]: httpfetch_init: parallel_limit=8
2018-06-29 10:00:01: INFO[Main]: Message locale is now set to: C
2018-06-29 10:00:01: VERBOSE[Main]: Determining world path
2018-06-29 10:00:01: INFO[Main]: Searching worlds...
2018-06-29 10:00:01: INFO[Main]:   In /storage/emulated/0/Minetest/worlds: 
2018-06-29 10:00:01: INFO[Main]: singleplayerworld 
2018-06-29 10:00:01: INFO[Main]: 1 found.
2018-06-29 10:00:01: [Main]: Automatically selecting world at [/storage/emulated/0/Minetest/worlds/singleplayerworld]
2018-06-29 10:00:01: VERBOSE[Main]: Determining gameid/gamespec
2018-06-29 10:00:01: INFO[Main]: Using world gameid [minetest]
2018-06-29 10:00:01: INFO[Main]: Using commanded world path [/storage/emulated/0/Minetest/worlds/singleplayerworld]
2018-06-29 10:00:01: INFO[Main]: Irrlicht: Irrlicht Engine version 1.9.0
2018-06-29 10:00:01: VERBOSE[Main]: Irrlicht: Waiting for Android activity window to be created.
2018-06-29 10:00:01: VERBOSE[Main]: Irrlicht: Android command APP_CMD_RESUME
2018-06-29 10:00:01: VERBOSE[Main]: Irrlicht: Android command APP_CMD_PAUSE
2018-06-29 10:00:01: VERBOSE[Main]: Irrlicht: Android command APP_CMD_INIT_WINDOW
2018-06-29 10:00:01: INFO[Main]: Irrlicht: EGL version: 1.400000
2018-06-29 10:00:01: VERBOSE[Main]: Irrlicht: EGL context created with OpenGLESVersion: : 1
2018-06-29 10:00:01: INFO[Main]: Irrlicht: Using renderer: OpenGL ES-CM 1.1 v1.r12p1-04bet0.1bb9662be2ebee934dcbd7265c794a91
2018-06-29 10:00:01: INFO[Main]: Irrlicht: ARM
2018-06-29 10:00:01: INFO[Main]: Irrlicht: GL_EXT_debug_marker GL_OES_byte_coordinates GL_OES_fixed_point GL_OES_single_precision GL_OES_matrix_get GL_OES_compressed_paletted_texture GL_OES_point_size_array GL_OES_point_sprite GL_OES_read_format GL_OES_compressed_ETC1_RGB8_texture GL_OES_depth24 GL_OES_stencil8 GL_OES_framebuffer_object GL_OES_packed_depth_stencil GL_OES_rgb8_rgba8 GL_EXT_read_format_bgra GL_OES_matrix_palette GL_OES_extended_matrix_palette GL_OES_draw_texture GL_OES_blend_equation_separate GL_OES_blend_func_separate GL_OES_blend_subtract GL_OES_stencil_wrap GL_OES_texture_mirrored_repeat GL_EXT_texture_format_BGRA8888 GL_OES_query_matrix GL_OES_EGL_image GL_OES_EGL_image_external GL_OES_EGL_sync GL_OES_texture_npot GL_OES_vertex_half_float GL_OES_required_internalformat GL_OES_vertex_array_object GL_OES_mapbuffer GL_OES_fbo_render_mipmap GL_OES_element_index_uint GL_ARM_rgba8 GL_EXT_blend_minmax GL_EXT_discard_framebuffer GL_EXT_texture_storage GL_OES_texture_compression_astc GL_KHR_texture_compression_astc_ldr GL_KHR_texture_compression_astc_hdr GL_KHR_texture_compression_astc_sliced_3d GL_OES_surfaceless_context GL_EXT_multisampled_render_to_texture GL_OES_texture_cube_map GL_KHR_debug GL_EXT_sRGB GL_EXT_robustness 
2018-06-29 10:00:02: VERBOSE[Main]: Irrlicht: Android command APP_CMD_RESUME
2018-06-29 10:00:02: VERBOSE[Main]: Irrlicht: Android command APP_CMD_GAINED_FOCUS
2018-06-29 10:00:02: INFO[Main]: text_height=35
2018-06-29 10:00:02: INFO[Main]: Waiting for other menus
2018-06-29 10:00:02: INFO[Main]: Waited for other menus
2018-06-29 10:00:02: INFO[Main]: Audio: Initializing...
2018-06-29 10:00:02: INFO[Main]: Audio: Initialized: OpenAL 1.1 ALSOFT 1.12.854, using Android Default
2018-06-29 10:00:02: INFO[Main]: GUIEngine: Initializing Lua
2018-06-29 10:00:02: VERBOSE[AsyncWorker-0]: Loading and running script from /storage/emulated/0/Minetest/builtin/init.lua
2018-06-29 10:00:02: VERBOSE[AsyncWorker-1]: Loading and running script from /storage/emulated/0/Minetest/builtin/init.lua
2018-06-29 10:00:02: VERBOSE[AsyncWorker-2]: Loading and running script from /storage/emulated/0/Minetest/builtin/init.lua
2018-06-29 10:00:02: VERBOSE[AsyncWorker-3]: Loading and running script from /storage/emulated/0/Minetest/builtin/init.lua
2018-06-29 10:00:02: INFO[Main]: SCRIPTAPI: Initialized main menu modules
2018-06-29 10:00:02: VERBOSE[Main]: Loading and running script from /storage/emulated/0/Minetest/builtin/init.lua
2018-06-29 10:00:02: INFO[AsyncWorker-2]: Initializing Asynchronous environment
2018-06-29 10:00:02: INFO[AsyncWorker-1]: Initializing Asynchronous environment
2018-06-29 10:00:02: INFO[AsyncWorker-0]: Initializing Asynchronous environment
2018-06-29 10:00:02: INFO[AsyncWorker-3]: Initializing Asynchronous environment
2018-06-29 10:00:02: INFO[Main]: Searching worlds...
2018-06-29 10:00:02: INFO[Main]:   In /storage/emulated/0/Minetest/worlds: 
2018-06-29 10:00:02: INFO[Main]: singleplayerworld 
2018-06-29 10:00:02: INFO[Main]: 1 found.
2018-06-29 10:00:02: VERBOSE[Main]: OpenALSoundManager::maintain(): 0 playing sounds, 0 sound names loaded
2018-06-29 10:00:02: INFO[Main]: Audio: Error opening /storage/emulated/0/Minetest/sounds/main_menu.0.ogg for decoding
2018-06-29 10:00:02: INFO[Main]: Audio: Error opening /storage/emulated/0/Minetest/sounds/main_menu.1.ogg for decoding
2018-06-29 10:00:02: INFO[Main]: Audio: Error opening /storage/emulated/0/Minetest/sounds/main_menu.2.ogg for decoding
2018-06-29 10:00:02: INFO[Main]: Audio: Error opening /storage/emulated/0/Minetest/sounds/main_menu.3.ogg for decoding
2018-06-29 10:00:02: INFO[Main]: Audio: Error opening /storage/emulated/0/Minetest/sounds/main_menu.4.ogg for decoding
2018-06-29 10:00:02: INFO[Main]: Audio: Error opening /storage/emulated/0/Minetest/sounds/main_menu.5.ogg for decoding
2018-06-29 10:00:02: INFO[Main]: Audio: Error opening /storage/emulated/0/Minetest/sounds/main_menu.6.ogg for decoding
2018-06-29 10:00:02: INFO[Main]: Audio: Error opening /storage/emulated/0/Minetest/sounds/main_menu.7.ogg for decoding
2018-06-29 10:00:02: INFO[Main]: Audio: Error opening /storage/emulated/0/Minetest/sounds/main_menu.8.ogg for decoding
2018-06-29 10:00:02: INFO[Main]: Audio: Error opening /storage/emulated/0/Minetest/sounds/main_menu.9.ogg for decoding
2018-06-29 10:00:02: INFO[Main]: Audio: Error opening /storage/emulated/0/Minetest/sounds/main_menu.ogg for decoding
2018-06-29 10:00:02: INFO[Main]: OpenALSoundManager: "main_menu" not found.
2018-06-29 10:00:07: INFO[Main]: Audio: Deinitializing...
2018-06-29 10:00:07: INFO[Main]: Audio: Deinitialized.
2018-06-29 10:00:07: INFO[Main]: GUIEngine: Deinitializing scripting
2018-06-29 10:00:07: WARNING[Main]: Irrlicht: Could not open file of texture: /storage/emulated/0/Minetest/textures/base/pack/server_flags_.png
2018-06-29 10:00:07: INFO[Main]: Searching worlds...
2018-06-29 10:00:07: INFO[Main]:   In /storage/emulated/0/Minetest/worlds: 
2018-06-29 10:00:07: INFO[Main]: singleplayerworld 
2018-06-29 10:00:07: INFO[Main]: 1 found.
2018-06-29 10:00:07: VERBOSE[Main]: httpfetch_cleanup: cleaning up
2018-06-29 10:00:07: INFO[Main]: Shutting down.
```
minetest.conf
```
screenH = 1080
screenW = 1776
selected_world_path = /storage/emulated/0/Minetest/worlds/singleplayerworld
main_menu_technical_settings = true
hud_scaling = 0.675
gui_scaling = 1
fast_move = false
invert_mouse = true
display_gamma = 3
show_debug = true
font_size = 10
mono_font_size = 10
fallback_font_size = 10
anisotropic_filter = false
bilinear_filter = false
fsaa = 0
leaves_style = fancy
mip_map = false
node_highlighting = box
touchscreen_threshold = 40
trilinear_filter = false
world_config_selected_mod = 1
ambient_occlusion_gamma = 4
free_move = false
noclip = false
movement_acceleration_fast = 5
movement_speed_fast = 10
debug_log_level = verbose
```
**Update:** logcat output captured while running minetest (only main menu): https://pastebin.com/0X378MUN
**Update2:** Solved by #7820, the local build made from latest commit on the mentioned PR fixes the issue.
##### Steps to reproduce

1. Install latest build from Google Play Store
2. Run it only entering main menu, enter and exit lock screen (same happens if ingame, so no difference)
3. See how game exits to phone interface, producing no crash log (clean exit)

**Comment by SmallJoker** [comment]
Should it pause instead? Minetest requires quite much power to run, so keeping it running normally is definitely not ideal.
Did this issue also exist in 0.4.16 and previous builds?

**Comment by juozaspo** [comment]
Had no such issue on 0.4.16, newer builds not tested, so this is new.

**Comment by ClobberXD** [comment]
This does not happen in APK v0.4.16 - the game pauses but does not shutdown.While that's not great for battery and temps, I think that's the expected behaviour. IMO, it's up to the user to keep MT running in the background, and battery drain / heating should be their own concern.

**Comment by red-001** [comment]
I think this is just the system closing the app for using too much memory.

**Comment by juozaspo** [comment]
On my phone the option to close memory intensive apps is turned off, and I doubt this is an issue, minetest isn't listed as memory intensive app too, nothing showing in closed memory intensive apps list. Also my phone has 2.0 GB of ram.

**Comment by nerzhul** [comment]
@ClobberXD i don't think. I remember sometimes i forgot to shutdown minetest and it eat my battery like a coward whereas i doesn't wanted that :)

**Comment by juozaspo** [comment]
Other apps like e.g. firefox doesn't close suddently like that. It's annoying when you accidently lock then unlock phone and find that the game window suddently goes away, it gets irritating after some time. It's very annoying to reopen minetest every time after it's gone. Worse when you're connected to a server, you have to retype the password when you open minetest before connecting.

**Comment by IhrFussel** [comment]
I got different results on my Android 7.0 phone:

When locking the screen in the main menu I get a black screen after unlocking. That black screen stays there until I go back to the home screen.

When locking the screen in SP game I get kicked back to the main menu after unlocking again.

**Comment by paramat** [comment]
As long as there isn't a bug here like black screen i think this is possibly desirable behaviour when locking the screen.

**Comment by juozaspo** [comment]
Some times when unlocking I see minetest "window" in wrong position, i.e. appearing cropped upright then going away instead of rotating back. Or wrongly rotated phone interface, or just a walpaper. logcat shows a focus event triggered in Irrlicht then game immediatelly exits.

**Comment by paramat** [comment]
Ok doesn't seem right then. No point running MT while screen is locked just to shut it down on unlock.

**Comment by juozaspo** [comment]
Done a test. Started a server on pc, connected Android client to in, locked screen, unlocked it. Disconnected only right after unlocking (still connected while screen is off and right before unlock)

**Comment by codexp** [comment]
On S9+ screen auto rotation causes game crash.

**Comment by paramat** [comment]
Could you open a new issue for that?

**Comment by juozaspo** [comment]
@codexp are you sure it's a crash and not a clean exit? If it's a crash you could produce backtrace using logcat command by running it inside adb shell or by using terminal emulator app and adding it to new issue. Also check your debug.txt too, try verbose output method like I used above.

**Comment by paramat** [comment]
Possibly fixed by #7820 please test.

**Comment by juozaspo** [comment]
I need to make a new build from latest master. I guess this might be fixed, not closing until confirmed.

**Comment by juozaspo** [comment]
Got build sorted. Needed to edit a few files. Patch (git diff), if any1 needs a quick test, same as in PR #7934:
```
diff --git a/src/gui/touchscreengui.cpp b/src/gui/touchscreengui.cpp
index a2d62c8..94e331f 100644
--- a/src/gui/touchscreengui.cpp
+++ b/src/gui/touchscreengui.cpp
@@ -22,12 +22,12 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "irrlichttypes.h"
 #include "irr_v2d.h"
 #include "log.h"
-#include "keycode.h"
+#include "client/keycode.h"
 #include "settings.h"
 #include "gettime.h"
 #include "util/numeric.h"
 #include "porting.h"
-#include "guiscalingfilter.h"
+#include "client/guiscalingfilter.h"
 
 #include <iostream>
 #include <algorithm>
diff --git a/src/gui/touchscreengui.h b/src/gui/touchscreengui.h
index 5102f18..f60051c 100644
--- a/src/gui/touchscreengui.h
+++ b/src/gui/touchscreengui.h
@@ -27,7 +27,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include <vector>
 
 #include "client/tile.h"
-#include "game.h"
+#include "client/game.h"
 
 using namespace irr;
 using namespace irr::core;
```

**Comment by juozaspo** [comment]
Confirmed fix from #7820 works, along with a patch from #7934, the issue were introduced by moving some files to client subdirectory of src subfolder. Closing this.
